File: concurrency/basic-fork-exa/1-single-fork-and-print.c
/* 
   Esempio super minimale di fork. Una sola fork, e sia il figlio che
   il padre non fanno altro che scrivere i due ID. Il padre non
   aspetta il figlio e quindi, se il figlio "ritarda" diventa uno
   'zombie' e viene 'adottato' dal processo 'init', per cui il parent
   ID che stampa non è quello del padre originale. Le due opzioni -c e
   -p permettono di scegliere di quanto "ritardare" l'esecuzione della
   printf nel figlio e nel padre; il default è zero per entrambi.
*/

#include <stdio.h>              /* printf */
#include <stdlib.h>             /* exit */
#include <unistd.h>             /* fork, getopt */
#include "fun.h"

int main(int argc, char *argv[])
{
  pid_t pid;
  int sleep_child = 0;
  int sleep_parent = 0;
  int opt;

  get_opt(argc, argv, 
          &sleep_child, &sleep_parent, 
          NULL);

  pid = fork();
  if (pid == 0) {
    
    sleep(sleep_child);
    printf("child:  %d with parent: %d\n", 
           getpid(), getppid());
    return 0;
  } else {
    
    sleep(sleep_parent);
    printf("parent: %d with child:  %d\n", 
           getpid(), pid);
    return 0;
  }
}

File: concurrency/basic-fork-exa/1-single-fork-and-print.py
# -*- coding:utf-8 -*-

# Esempio super minimale di fork. Una sola fork, e sia il figlio che
# il padre non fanno altro che scrivere i due ID. Il padre non aspetta
# il figlio e quindi, se il figlio "ritarda" diventa uno 'zombie' e
# viene 'adottato' dal processo 'init', per cui il parent ID che
# stampa non è quello del padre originale. Le due opzioni -c e -p
# permettono di scegliere di quanto "ritardare" l'esecuzione della
# printf nel figlio e nel padre; il default è zero per entrambi.

import os
import time
import random
import sys
import argparse

def show(s):
    sys.stderr.write(s + "\n")
    
def go(child_sleep, parent_sleep):
    pid = os.fork()
    if pid == 0:

        time.sleep(child_sleep)
        show("child:  %d with parent: %d" % (
            os.getpid(), os.getppid()))
        sys.exit(0)        
    else:  
        time.sleep(parent_sleep)

def parse_args(args):

    parser = argparse.ArgumentParser()
    add = parser.add_argument("-c", "--child-sleep", help="Child max sleep time (seconds).", type=float, default=0.5)
    add = parser.add_argument("-p", "--parent-sleep", help="Parent max sleep time (seconds).", type=float, default=3.0)
    opt = parser.parse_args(args)
    return opt

def main(args):

    opt = parse_args(args)
    show ("Runner: %d" % os.getpid())
    for _ in range(5):
        go(random.random() * opt.child_sleep,
           random.random() * opt.parent_sleep)

if __name__ == "__main__":

    import sys
    main(sys.argv[1:])
